File: lib/python/genesis/core/alma_memory.py
import json
import re
from dataclasses import asdict, dataclass
from typing import Any, Dict, List, Optional


WORD_PATTERN = re.compile(r"[a-zA-Z_][a-zA-Z0-9_]{2,}")


@dataclass
class ALMAMemoryEntry:
    generation: int
    memory_type: str  # success | failure
    patch_type: str
    patch_name: str
    patch_description: str
    score_delta: float
    combined_score: float
    summary: str
    keywords: List[str]


class ALMAMemorySystem:
    """ALMA-inspired long-term memory for evolutionary code agents.

    This stores structured mutation outcomes and retrieves relevant, concise
    memories to ground future patch generation.
    """

    def __init__(
        self,
        enabled: bool = False,
        max_entries: int = 256,
        max_retrievals: int = 4,
        min_success_delta: float = 0.0,
    ):
        self.enabled = enabled
        self.max_entries = max(1, max_entries)
        self.max_retrievals = max(1, max_retrievals)
        self.min_success_delta = min_success_delta
        self.entries: List[ALMAMemoryEntry] = []

    def observe_outcome(
        self,
        generation: int,
        parent_score: float,
        child_score: float,
        correct: bool,
        patch_type: Optional[str],
        patch_name: Optional[str],
        patch_description: Optional[str],
        diff_summary: Any,
        text_feedback: Optional[str],
        error_message: Optional[str],
    ) -> None:
        """Capture one mutation episode as durable memory."""
        if not self.enabled:
            return

        score_delta = child_score - parent_score
        is_success = correct and score_delta >= self.min_success_delta
        memory_type = "success" if is_success else "failure"

        summary_parts: List[str] = []
        if patch_description:
            summary_parts.append(patch_description.strip())
        if diff_summary:
            summary_parts.append(self._compact_text(diff_summary, 180))
        if text_feedback:
            summary_parts.append(self._compact_text(text_feedback, 180))
        if error_message:
            summary_parts.append(self._compact_text(error_message, 180))
        summary = " | ".join(part for part in summary_parts if part)[:500]

        keywords = self._extract_keywords(
            " ".join(
                [
                    str(patch_type or ""),
                    str(patch_name or ""),
                    str(patch_description or ""),
                    str(diff_summary or ""),
                    str(text_feedback or ""),
                    str(error_message or ""),
                ]
            )
        )

        entry = ALMAMemoryEntry(
            generation=generation,
            memory_type=memory_type,
            patch_type=patch_type or "unknown",
            patch_name=patch_name or "",
            patch_description=patch_description or "",
            score_delta=score_delta,
            combined_score=child_score,
            summary=summary,
            keywords=keywords,
        )
        self.entries.append(entry)
        self._trim_entries()

    def build_prompt_context(
        self,
        current_generation: int,
        parent_code: str,
        parent_feedback: Optional[str],
    ) -> Optional[str]:
        """Retrieve memory snippets relevant to current parent context."""
        if not self.enabled or not self.entries:
            return None

        context_keywords = self._extract_keywords(
            f"{parent_code[:2000]} {str(parent_feedback or '')}"
        )
        scored: List[tuple[float, ALMAMemoryEntry]] = []
        for entry in self.entries:
            overlap = len(set(context_keywords) & set(entry.keywords))
            recency_bonus = 1.0 / (1.0 + max(0, current_generation - entry.generation))
            impact = abs(entry.score_delta)
            score = (2.0 * overlap) + recency_bonus + impact
            scored.append((score, entry))

        scored.sort(key=lambda x: x[0], reverse=True)
        selected = [entry for _, entry in scored[: self.max_retrievals]]
        if not selected:
            return None

        success_entries = [e for e in selected if e.memory_type == "success"]
        failure_entries = [e for e in selected if e.memory_type == "failure"]
        lines: List[str] = ["# ALMA Long-Term Memory"]
        if success_entries:
            lines.append("## Reuse These Successful Patterns")
            for item in success_entries:
                lines.append(
                    f"- {item.patch_type} `{item.patch_name or 'unnamed'}` "
                    f"(delta {item.score_delta:+.4f}): {item.summary or 'No summary.'}"
                )
        if failure_entries:
            lines.append("## Avoid These Failure Patterns")
            for item in failure_entries:
                lines.append(
                    f"- {item.patch_type} `{item.patch_name or 'unnamed'}` "
                    f"(delta {item.score_delta:+.4f}): {item.summary or 'No summary.'}"
                )
        return "\n".join(lines)

    def save_state(self, path: str) -> None:
        if not self.enabled:
            return
        payload = {
            "enabled": self.enabled,
            "max_entries": self.max_entries,
            "max_retrievals": self.max_retrievals,
            "min_success_delta": self.min_success_delta,
            "entries": [asdict(entry) for entry in self.entries],
        }
        with open(path, "w", encoding="utf-8") as f:
            json.dump(payload, f, indent=2)

    def load_state(self, path: str) -> bool:
        if not self.enabled:
            return False
        try:
            with open(path, "r", encoding="utf-8") as f:
                payload = json.load(f)
        except Exception:
            return False

        loaded_entries: List[ALMAMemoryEntry] = []
        for item in payload.get("entries", []):
            try:
                loaded_entries.append(ALMAMemoryEntry(**item))
            except TypeError:
                continue
        self.entries = loaded_entries
        self._trim_entries()
        return True

    def _trim_entries(self) -> None:
        if len(self.entries) <= self.max_entries:
            return
        self.entries.sort(
            key=lambda x: (abs(x.score_delta), x.generation),
            reverse=True,
        )
        self.entries = self.entries[: self.max_entries]

    def _extract_keywords(self, text: str) -> List[str]:
        if not text:
            return []
        words = [w.lower() for w in WORD_PATTERN.findall(text)]
        stopwords = {
            "the",
            "and",
            "for",
            "with",
            "that",
            "this",
            "from",
            "into",
            "code",
            "patch",
            "generation",
            "score",
            "delta",
            "result",
            "error",
        }
        unique = []
        seen = set()
        for word in words:
            if word in stopwords or word in seen:
                continue
            seen.add(word)
            unique.append(word)
        return unique[:30]

    def _compact_text(self, obj: Any, max_len: int) -> str:
        if obj is None:
            return ""
        text = str(obj).replace("\n", " ").strip()
        if len(text) <= max_len:
            return text
        return text[: max_len - 3] + "..."

File: genesis_rust_backend/src/core/alma_memory.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlmaMemoryEntry {
    pub generation: usize,
    pub memory_type: String,
    pub patch_type: String,
    pub patch_name: String,
    pub patch_description: String,
    pub score_delta: f64,
    pub combined_score: f64,
    pub summary: String,
    pub keywords: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlmaMemorySystem {
    pub enabled: bool,
    pub max_entries: usize,
    pub max_retrievals: usize,
    pub min_success_delta: f64,
    pub entries: Vec<AlmaMemoryEntry>,
}

impl AlmaMemorySystem {
    pub fn new(
        enabled: bool,
        max_entries: usize,
        max_retrievals: usize,
        min_success_delta: f64,
    ) -> Self {
        Self {
            enabled,
            max_entries: max_entries.max(1),
            max_retrievals: max_retrievals.max(1),
            min_success_delta,
            entries: Vec::new(),
        }
    }

    #[allow(clippy::too_many_arguments)]
    pub fn observe_outcome(
        &mut self,
        generation: usize,
        parent_score: f64,
        child_score: f64,
        correct: bool,
        patch_type: &str,
        patch_name: &str,
        patch_description: &str,
        diff_summary: &str,
        text_feedback: &str,
        error_message: &str,
    ) {
        if !self.enabled {
            return;
        }

        let score_delta = child_score - parent_score;
        let memory_type = if correct && score_delta >= self.min_success_delta {
            "success"
        } else {
            "failure"
        }
        .to_string();

        let summary = [
            patch_description,
            diff_summary,
            text_feedback,
            error_message,
        ]
        .iter()
        .filter(|s| !s.is_empty())
        .copied()
        .collect::<Vec<_>>()
        .join(" | ");

        let keywords = extract_keywords(&format!(
            "{} {} {} {} {} {}",
            patch_type, patch_name, patch_description, diff_summary, text_feedback, error_message
        ));

        self.entries.push(AlmaMemoryEntry {
            generation,
            memory_type,
            patch_type: patch_type.to_string(),
            patch_name: patch_name.to_string(),
            patch_description: patch_description.to_string(),
            score_delta,
            combined_score: child_score,
            summary,
            keywords,
        });

        self.entries
            .sort_by(|a, b| b.score_delta.abs().total_cmp(&a.score_delta.abs()));
        if self.entries.len() > self.max_entries {
            self.entries.truncate(self.max_entries);
        }
    }

    pub fn build_prompt_context(
        &self,
        current_generation: usize,
        parent_code: &str,
        parent_feedback: &str,
    ) -> Option<String> {
        if !self.enabled || self.entries.is_empty() {
            return None;
        }

        let context_keywords = extract_keywords(&format!("{} {}", parent_code, parent_feedback));
        let context_set: HashSet<String> = context_keywords.into_iter().collect();

        let mut scored = self
            .entries
            .iter()
            .map(|entry| {
                let overlap = entry
                    .keywords
                    .iter()
                    .filter(|k| context_set.contains(*k))
                    .count() as f64;
                let recency =
                    1.0 / (1.0 + current_generation.saturating_sub(entry.generation) as f64);
                let impact = entry.score_delta.abs();
                (2.0 * overlap + recency + impact, entry)
            })
            .collect::<Vec<_>>();

        scored.sort_by(|a, b| b.0.total_cmp(&a.0));
        let selected = scored
            .into_iter()
            .take(self.max_retrievals)
            .map(|(_, entry)| entry)
            .collect::<Vec<_>>();

        if selected.is_empty() {
            return None;
        }

        let mut out = String::from("# ALMA Long-Term Memory\n");
        let successes = selected
            .iter()
            .filter(|e| e.memory_type == "success")
            .collect::<Vec<_>>();
        let failures = selected
            .iter()
            .filter(|e| e.memory_type == "failure")
            .collect::<Vec<_>>();

        if !successes.is_empty() {
            out.push_str("## Reuse These Successful Patterns\n");
            for s in successes {
                out.push_str(&format!(
                    "- {} `{}` (delta {:+.4}): {}\n",
                    s.patch_type, s.patch_name, s.score_delta, s.summary
                ));
            }
        }

        if !failures.is_empty() {
            out.push_str("## Avoid These Failure Patterns\n");
            for f in failures {
                out.push_str(&format!(
                    "- {} `{}` (delta {:+.4}): {}\n",
                    f.patch_type, f.patch_name, f.score_delta, f.summary
                ));
            }
        }

        Some(out)
    }

    pub fn save_state(&self, path: &str) -> anyhow::Result<()> {
        if !self.enabled {
            return Ok(());
        }
        let payload = serde_json::to_string_pretty(self)?;
        std::fs::write(path, payload)?;
        Ok(())
    }

    pub fn load_state(&mut self, path: &str) -> bool {
        if !self.enabled {
            return false;
        }
        let Ok(raw) = std::fs::read_to_string(path) else {
            return false;
        };
        let Ok(parsed) = serde_json::from_str::<AlmaMemorySystem>(&raw) else {
            return false;
        };
        self.entries = parsed.entries;
        true
    }
}

fn extract_keywords(text: &str) -> Vec<String> {
    let stopwords: HashSet<&str> = [
        "the",
        "and",
        "for",
        "with",
        "that",
        "this",
        "from",
        "into",
        "code",
        "patch",
        "generation",
        "score",
        "delta",
        "result",
        "error",
    ]
    .iter()
    .copied()
    .collect();

    let mut seen = HashSet::new();
    let mut out = Vec::new();
    for token in text
        .split(|c: char| !c.is_alphanumeric() && c != '_')
        .map(|t| t.to_lowercase())
    {
        if token.len() < 3 || stopwords.contains(token.as_str()) || seen.contains(&token) {
            continue;
        }
        seen.insert(token.clone());
        out.push(token);
        if out.len() >= 30 {
            break;
        }
    }
    out
}
